src/nxt_unit_typetraits.h

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2022-2024, Alejandro Colomar <alx@kernel.org>
+ */
+
+#ifndef _NXT_UNIT_TYPETRAITS_H_INCLUDED_
+#define _NXT_UNIT_TYPETRAITS_H_INCLUDED_
+
+
+#include <limits.h>
+
+
+#define nxt_widthof(T)      (sizeof(T) * CHAR_BIT)
+
+
+#define nxt_is_unsigned(x)  (((typeof(x)) -1) > 1)
+#define nxt_is_signed(x)    (((typeof(x)) -1) < 1)
+
+
+#define nxt_stype_max(T)                                                      \
```

Comment:
We can't overflow, so this calculation is tricky.

Let's say our type has 8 bits.

```c
(widthof(T) - 2)  // 6
```

```c
((T) 1 << 6)   // 0b0100_0000
```

```c
((0b0100_0000 - 1) << 1)  // 0x0111_1110
```

```c
(0b0111_1110 + 1)  // 0b0111_1111
```

Works with any signed integer without triggering UB

Reply:
See what glibc has:

```c
$ grep -rn -- '- *2)* *- *1)'
atoi/time/tst-mktime2.c:26:	: ((((t) 1 << (sizeof (t) * CHAR_BIT - 2)) - 1) * 2 + 1)))
atoi/include/intprops.h:59:        : ((((t) 1 << (TYPE_WIDTH (t) - 2)) - 1) * 2 + 1)))
atoi/include/intprops.h:72:  (((_GL_INT_CONVERT (e, 1) << (TYPE_WIDTH (+ (e)) - 2)) - 1) * 2 + 1)
null/time/tst-mktime2.c:26:	: ((((t) 1 << (sizeof (t) * CHAR_BIT - 2)) - 1) * 2 + 1)))
null/include/intprops.h:59:        : ((((t) 1 << (TYPE_WIDTH (t) - 2)) - 1) * 2 + 1)))
null/include/intprops.h:72:  (((_GL_INT_CONVERT (e, 1) << (TYPE_WIDTH (+ (e)) - 2)) - 1) * 2 + 1)
sscanf/time/tst-mktime2.c:26:	: ((((t) 1 << (sizeof (t) * CHAR_BIT - 2)) - 1) * 2 + 1)))
sscanf/include/intprops.h:59:        : ((((t) 1 << (TYPE_WIDTH (t) - 2)) - 1) * 2 + 1)))
sscanf/include/intprops.h:72:  (((_GL_INT_CONVERT (e, 1) << (TYPE_WIDTH (+ (e)) - 2)) - 1) * 2 + 1)
master/time/tst-mktime2.c:26:	: ((((t) 1 << (sizeof (t) * CHAR_BIT - 2)) - 1) * 2 + 1)))
master/include/intprops.h:59:        : ((((t) 1 << (TYPE_WIDTH (t) - 2)) - 1) * 2 + 1)))
master/include/intprops.h:72:  (((_GL_INT_CONVERT (e, 1) << (TYPE_WIDTH (+ (e)) - 2)) - 1) * 2 + 1)
```